# RFM ANALYSIS FOR ONLİNE RETAIL DATA SET

## Data Set Information:

https://archive.ics.uci.edu/ml/datasets/Online+Retail+II

This Online Retail II data set contains all the transactions occurring for a UK-based and registered, non-store online retail between 01/12/2009 and 09/12/2011.The company mainly sells unique all-occasion gift-ware. Many customers of the company are wholesalers.

## Attribute Information:

InvoiceNo: Invoice number. Nominal. A 6-digit integral number uniquely assigned to each transaction. If this code starts with the letter 'c', it indicates a cancellation.
StockCode: Product (item) code. Nominal. A 5-digit integral number uniquely assigned to each distinct product.
Description: Product (item) name. Nominal.
Quantity: The quantities of each product (item) per transaction. Numeric.
InvoiceDate: Invice date and time. Numeric. The day and time when a transaction was generated.
UnitPrice: Unit price. Numeric. Product price per unit in sterling (£).
CustomerID: Customer number. Nominal. A 5-digit integral number uniquely assigned to each customer.
Country: Country name. Nominal. The name of the country where a customer resides.

## Data Understanding

In [106]:
import pandas as pd
import numpy as np
import seaborn as sns

In [107]:
pd.set_option('display.max_columns', None); pd.set_option('display.max_rows', None);

pd.set_option('display.float_format', lambda x: '%.0f' % x)
import matplotlib.pyplot as plt

In [91]:
df_2010_2011 = pd.read_excel("online_retail_II.xlsx", sheet_name = "Year 2010-2011")

In [92]:
df = df_2010_2011.copy()

In [109]:
df.head()

Invoice StockCode                          Description  Quantity  \
0  536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1  536365     71053                  WHITE METAL LANTERN         6   
2  536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3  536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4  536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  Price  Customer ID         Country  
0 2010-12-01 08:26:00      3        17850  United Kingdom  
1 2010-12-01 08:26:00      3        17850  United Kingdom  
2 2010-12-01 08:26:00      3        17850  United Kingdom  
3 2010-12-01 08:26:00      3        17850  United Kingdom  
4 2010-12-01 08:26:00      3        17850  United Kingdom

In [110]:
# Unique products
df["Description"].nunique()

4223

In [111]:
# Each products counts are..
df["Description"].value_counts().head()

WHITE HANGING HEART T-LIGHT HOLDER    2369
REGENCY CAKESTAND 3 TIER              2200
JUMBO BAG RED RETROSPOT               2159
PARTY BUNTING                         1727
LUNCH BAG RED RETROSPOT               1638
Name: Description, dtype: int64

In [112]:
# Best-seller
df.groupby("Description").agg({"Quantity":"sum"}).head()

Quantity
Description                             
20713                               -400
 4 PURPLE FLOCK DINNER CANDLES       144
 50'S CHRISTMAS GIFT BAG LARGE      1913
 DOLLY GIRL BEAKER                  2448
 I LOVE LONDON MINI BACKPACK         389

In [113]:
df.groupby("Description").agg({"Quantity":"sum"}).sort_values("Quantity", ascending = False).head()

Quantity
Description                                
WORLD WAR 2 GLIDERS ASSTD DESIGNS     53847
JUMBO BAG RED RETROSPOT               47363
ASSORTED COLOUR BIRD ORNAMENT         36381
POPCORN HOLDER                        36334
PACK OF 72 RETROSPOT CAKE CASES       36039

In [114]:
# Unique invoice
df["Invoice"].nunique()

25900

In [115]:
df["TotalPrice"] = df["Quantity"]*df["Price"]

In [116]:
df.head()

Invoice StockCode                          Description  Quantity  \
0  536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1  536365     71053                  WHITE METAL LANTERN         6   
2  536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3  536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4  536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  Price  Customer ID         Country  TotalPrice  
0 2010-12-01 08:26:00      3        17850  United Kingdom          15  
1 2010-12-01 08:26:00      3        17850  United Kingdom          20  
2 2010-12-01 08:26:00      3        17850  United Kingdom          22  
3 2010-12-01 08:26:00      3        17850  United Kingdom          20  
4 2010-12-01 08:26:00      3        17850  United Kingdom          20

In [117]:
# The top invoices for price
df.groupby("Invoice").agg({"TotalPrice":"sum"}).head()

TotalPrice
Invoice            
536365          139
536366           22
536367          279
536368           70
536369           18

In [118]:
# The most expensive product is "POSTAGE"
df.sort_values("Price", ascending = False).head()

Invoice  StockCode Description  Quantity         InvoiceDate  Price  \
222681  C556445          M      Manual        -1 2011-06-10 15:31:00  38970   
524602  C580605  AMAZONFEE  AMAZON FEE        -1 2011-12-05 11:36:00  17836   
43702   C540117  AMAZONFEE  AMAZON FEE        -1 2011-01-05 09:55:00  16888   
43703   C540118  AMAZONFEE  AMAZON FEE        -1 2011-01-05 09:57:00  16454   
16356   C537651  AMAZONFEE  AMAZON FEE        -1 2010-12-07 15:49:00  13541   

        Customer ID         Country  TotalPrice  
222681        15098  United Kingdom      -38970  
524602          nan  United Kingdom      -17836  
43702           nan  United Kingdom      -16888  
43703           nan  United Kingdom      -16454  
16356           nan  United Kingdom      -13541

In [119]:
df["Country"].value_counts().head()

United Kingdom    495478
Germany             9495
France              8558
EIRE                8196
Spain               2533
Name: Country, dtype: int64

Data Preparation

In [120]:
df.isnull().sum()

Invoice             0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
Price               0
Customer ID    135080
Country             0
TotalPrice          0
dtype: int64

In [121]:
df.dropna(inplace = True)

In [122]:
df.shape

(406830, 9)

In [123]:
df.describe([0.01,0.05,0.10,0.25,0.50,0.75,0.90,0.95, 0.99]).T

count  mean  std     min    1%    5%   10%   25%   50%   75%  \
Quantity    406830    12  249  -80995    -2     1     1     2     5    12   
Price       406830     3   69       0     0     0     1     1     2     4   
Customer ID 406830 15288 1714   12346 12415 12626 12876 13953 15152 16791   
TotalPrice  406830    20  428 -168470   -10     1     2     4    11    20   

              90%   95%   99%    max  
Quantity       24    36   120  80995  
Price           7     8    15  38970  
Customer ID 17719 17905 18212  18287  
TotalPrice     35    67   200 168470

In [124]:
for feature in ["Quantity","Price","TotalPrice"]:

    Q1 = df[feature].quantile(0.01)
    Q3 = df[feature].quantile(0.99)
    IQR = Q3-Q1
    upper = Q3 + 1.5*IQR
    lower = Q1 - 1.5*IQR

    if df[(df[feature] > upper) | (df[feature] < lower)].any(axis=None):
        print(feature,"yes")
        print(df[(df[feature] > upper) | (df[feature] < lower)].shape[0])
    else:
        print(feature, "no")

Quantity yes
948
Price yes
846
TotalPrice yes
1030


## RFM SCORES

In [125]:
df.head()

Invoice StockCode                          Description  Quantity  \
0  536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1  536365     71053                  WHITE METAL LANTERN         6   
2  536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3  536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4  536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  Price  Customer ID         Country  TotalPrice  
0 2010-12-01 08:26:00      3        17850  United Kingdom          15  
1 2010-12-01 08:26:00      3        17850  United Kingdom          20  
2 2010-12-01 08:26:00      3        17850  United Kingdom          22  
3 2010-12-01 08:26:00      3        17850  United Kingdom          20  
4 2010-12-01 08:26:00      3        17850  United Kingdom          20

### Recency

In [126]:
df["InvoiceDate"].min()

Timestamp('2010-12-01 08:26:00')

In [127]:
df["InvoiceDate"].max()

Timestamp('2011-12-09 12:50:00')

In [128]:
import datetime as dt
today_date = dt.datetime(2011,12,9)

In [129]:
today_date

datetime.datetime(2011, 12, 9, 0, 0)

In [130]:
df.groupby("Customer ID").agg({"InvoiceDate":"max"}).head()

InvoiceDate
Customer ID                    
12346       2011-01-18 10:17:00
12347       2011-12-07 15:52:00
12348       2011-09-25 13:13:00
12349       2011-11-21 09:51:00
12350       2011-02-02 16:01:00

In [131]:
df["Customer ID"] = df["Customer ID"].astype(int)

In [132]:
(today_date - df.groupby("Customer ID").agg({"InvoiceDate":"max"})).head()

InvoiceDate
Customer ID                  
12346       324 days 13:43:00
12347         1 days 08:08:00
12348        74 days 10:47:00
12349        17 days 14:09:00
12350       309 days 07:59:00

In [133]:
temp_df = (today_date - df.groupby("Customer ID").agg({"InvoiceDate":"max"}))

In [134]:
temp_df.rename(columns={"InvoiceDate": "Recency"}, inplace = True)

In [135]:
temp_df.head()

Recency
Customer ID                  
12346       324 days 13:43:00
12347         1 days 08:08:00
12348        74 days 10:47:00
12349        17 days 14:09:00
12350       309 days 07:59:00

In [142]:
recency_df = temp_df["Recency"].apply(lambda x: x.days)

In [143]:
recency_df.head()

Customer ID
12346    324
12347      1
12348     74
12349     17
12350    309
Name: Recency, dtype: int64

### Frequency

In [145]:
temp_df = df.groupby(["Customer ID","Invoice"]).agg({"Invoice":"count"})

In [146]:
temp_df.head()

Invoice
Customer ID Invoice         
12346       541431         1
            C541433        1
12347       537626        31
            542237        29
            549222        24

In [147]:
temp_df.groupby("Customer ID").agg({"Invoice":"sum"}).head()

Invoice
Customer ID         
12346              2
12347            182
12348             31
12349             73
12350             17

In [148]:
freq_df = temp_df.groupby("Customer ID").agg({"Invoice":"sum"})
freq_df.rename(columns={"Invoice": "Frequency"}, inplace = True)
freq_df.head()

Frequency
Customer ID           
12346                2
12347              182
12348               31
12349               73
12350               17

### Monetary

In [149]:
monetary_df = df.groupby("Customer ID").agg({"TotalPrice":"sum"})

In [150]:
monetary_df.head()

TotalPrice
Customer ID            
12346                 0
12347              4310
12348              1797
12349              1758
12350               334

In [151]:
monetary_df.rename(columns={"TotalPrice": "Monetary"}, inplace = True)

In [152]:
print(recency_df.shape,freq_df.shape,monetary_df.shape)

(4372,) (4372, 1) (4372, 1)


In [153]:
rfm = pd.concat([recency_df, freq_df, monetary_df],  axis=1)

In [154]:
rfm.head()

Recency  Frequency  Monetary
Customer ID                              
12346            324          2         0
12347              1        182      4310
12348             74         31      1797
12349             17         73      1758
12350            309         17       334

In [155]:
rfm["RecencyScore"] = pd.qcut(rfm['Recency'], 5, labels = [5, 4, 3, 2, 1])

In [156]:
rfm["FrequencyScore"] = pd.qcut(rfm['Frequency'], 5, labels = [1, 2, 3, 4, 5])

In [157]:
rfm["MonetaryScore"] = pd.qcut(rfm['Monetary'], 5, labels = [1, 2, 3, 4, 5])

In [158]:
rfm.head()

Recency  Frequency  Monetary RecencyScore FrequencyScore  \
Customer ID                                                             
12346            324          2         0            1              1   
12347              1        182      4310            5              5   
12348             74         31      1797            2              3   
12349             17         73      1758            4              4   
12350            309         17       334            1              2   

            MonetaryScore  
Customer ID                
12346                   1  
12347                   5  
12348                   4  
12349                   4  
12350                   2

In [159]:
(rfm['RecencyScore'].astype(str) + 
 rfm['FrequencyScore'].astype(str) + 
 rfm['MonetaryScore'].astype(str)).head()

Customer ID
12346    111
12347    555
12348    234
12349    444
12350    122
dtype: object

In [160]:
rfm["RFM_SCORE"] = rfm['RecencyScore'].astype(str) + rfm['FrequencyScore'].astype(str) + rfm['MonetaryScore'].astype(str)

In [161]:
rfm.head()

Recency  Frequency  Monetary RecencyScore FrequencyScore  \
Customer ID                                                             
12346            324          2         0            1              1   
12347              1        182      4310            5              5   
12348             74         31      1797            2              3   
12349             17         73      1758            4              4   
12350            309         17       334            1              2   

            MonetaryScore RFM_SCORE  
Customer ID                          
12346                   1       111  
12347                   5       555  
12348                   4       234  
12349                   4       444  
12350                   2       122

In [162]:
rfm.describe().T

count  mean  std   min  25%  50%  75%    max
Recency     4372    91  101    -1   15   49  142    372
Frequency   4372    93  232     1   17   42  102   7983
Monetary    4372  1898 8219 -4288  293  648 1612 279489

In [163]:
rfm[rfm["RFM_SCORE"] == "555"].head()

Recency  Frequency  Monetary RecencyScore FrequencyScore  \
Customer ID                                                             
12347              1        182      4310            5              5   
12359              6        254      6246            5              5   
12362              2        274      5155            5              5   
12417              2        198      3579            5              5   
12433             -1        420     13376            5              5   

            MonetaryScore RFM_SCORE  
Customer ID                          
12347                   5       555  
12359                   5       555  
12362                   5       555  
12417                   5       555  
12433                   5       555

In [164]:
rfm[rfm["RFM_SCORE"] == "111"].head()

Recency  Frequency  Monetary RecencyScore FrequencyScore  \
Customer ID                                                             
12346            324          2         0            1              1   
12353            203          4        89            1              1   
12361            286         10       190            1              1   
12401            302          5        84            1              1   
12402            322         11       226            1              1   

            MonetaryScore RFM_SCORE  
Customer ID                          
12346                   1       111  
12353                   1       111  
12361                   1       111  
12401                   1       111  
12402                   1       111

In [166]:
seg_map = {
    r'[1-2][1-2]': 'Hibernating',
    r'[1-2][3-4]': 'At Risk',
    r'[1-2]5': 'Can\'t Loose',
    r'3[1-2]': 'About to Sleep',
    r'33': 'Need Attention',
    r'[3-4][4-5]': 'Loyal Customers',
    r'41': 'Promising',
    r'51': 'New Customers',
    r'[4-5][2-3]': 'Potential Loyalists',
    r'5[4-5]': 'Champions'
}

In [167]:
rfm['Segment'] = rfm['RecencyScore'].astype(str) + rfm['FrequencyScore'].astype(str)
rfm['Segment'] = rfm['Segment'].replace(seg_map, regex=True)
rfm.head()

Recency  Frequency  Monetary RecencyScore FrequencyScore  \
Customer ID                                                             
12346            324          2         0            1              1   
12347              1        182      4310            5              5   
12348             74         31      1797            2              3   
12349             17         73      1758            4              4   
12350            309         17       334            1              2   

            MonetaryScore RFM_SCORE          Segment  
Customer ID                                           
12346                   1       111      Hibernating  
12347                   5       555        Champions  
12348                   4       234          At Risk  
12349                   4       444  Loyal Customers  
12350                   2       122      Hibernating

In [168]:
rfm[["Segment", "Recency","Frequency","Monetary"]].groupby("Segment").agg(["mean","count"])

Recency       Frequency       Monetary      
                       mean count      mean count     mean count
Segment                                                         
About to Sleep           51   328        16   328      412   328
At Risk                 164   577        57   577      950   577
Can't Loose             141    81       184    81     2346    81
Champions                 4   620       287   620     6788   620
Hibernating             211  1066        13  1066      369  1066
Loyal Customers          31   825       162   825     2581   825
Need Attention           50   208        42   208      833   208
New Customers             5    60         8    60      660    60
Potential Loyalists      14   502        35   502      877   502
Promising                21   105         8   105      413   105

## Evaluation and Action Proposal

Within this table, information about 3 customer groups were analyzed and actions related to these analyzes were determined.

1. Cant Loose Them: This group consists of 81 customers. The Recency average is 141, the Frequency average is 184, and the Monetary average is 2346.

2. Horse Risk: This group consists of 577 customers. Recency averages are 164, Frequency averages are 57, and Monetary averages are 950.

3. Need Attention: This group consists of 208 customers. Recency averages are 50, Frequency averages are 42, Monetary averages are 833.

Yorumlar

The most risky and potentially focused customers on this table are the "Can't Loose Them" class. This class is the most valuable customer class in this table. Recency value is low, but it is the highest when looking at Frequence. In other words, this customer is the class with the highest potential to leave us. Considering all customer classes, Monetary is one of the customers with the highest value, but this group of customers has not recently exchanged. He approached leaving our company. This is the most important customer class for our company.

A class that should not be lost is the "Horse Risk" group of customers. This group is also one of the highest classes in Frequency. It also attracts attention as it is the third-class customer group that is the most customer. Recency value of this group is low. This is an indication that this group has not been shopping recently. It is necessary to bring these customers back to the company.

In the "Need Attention" class, frequency and recency values ​​are an average customer group. This customer group is also a class that needs attention because it has the potential to shift to both regions. As an action suggestion, if supported by promotions, discounts and loyalty to the company, this customer group can be shifted towards loyal customers or potential loyal customers. However, as a result of such action attempts, they may also shift to the "horse risk" group or they may also be included in the "about to sleep" class.

As actions to be taken for these groups, their loyalty should be increased through promotions, discounts and special team campaigns. Special activities should be carried out for these customer groups. Actions to be taken for these groups will provide higher efficiency compared to other customer classes. For this reason, information about these 3 customer groups should be extracted and this customer information should be shared with the business intelligence department. These departments, which have a business intelligence department, should be directed to departments such as purchasing and marketing by removing the contact information of these customers and they should mobilize these customers regarding their own business areas.

Below, information about the "Need Attention" customer group has been extracted as "new_customers.csv" file. This file is prepared to be sent to the business intelligence department.

In [169]:
rfm[rfm["Segment"] == "Need Attention"].head()

Recency  Frequency  Monetary RecencyScore FrequencyScore  \
Customer ID                                                             
12372             70         52      1298            3              3   
12413             65         40       694            3              3   
12446             56         58      1002            3              3   
12454             52         30        -0            3              3   
12458             70         38       947            3              3   

            MonetaryScore RFM_SCORE         Segment  
Customer ID                                          
12372                   4       334  Need Attention  
12413                   3       333  Need Attention  
12446                   4       334  Need Attention  
12454                   1       331  Need Attention  
12458                   4       334  Need Attention

In [170]:
rfm[rfm["Segment"] == "New Customers"].index

Int64Index([12367, 12442, 12587, 12660, 12798, 12882, 12966, 13017, 13068,
            13079, 13147, 13188, 13255, 13298, 13404, 13436, 14219, 14287,
            14354, 14385, 14480, 14520, 14569, 14601, 14777, 14785, 14804,
            14836, 14853, 14860, 14865, 15101, 15195, 15206, 15619, 15773,
            15783, 15793, 15992, 16000, 16446, 16500, 16569, 16597, 16620,
            16789, 16852, 16988, 16989, 17044, 17436, 17468, 17737, 17929,
            17942, 18030, 18058, 18174, 18273, 18282],
           dtype='int64', name='Customer ID')

In [171]:
new_df = pd.DataFrame()
new_df["NewCustomerID"] = rfm[rfm["Segment"] == "New Customers"].index

In [172]:
new_df.head()

NewCustomerID
0          12367
1          12442
2          12587
3          12660
4          12798

In [173]:
new_df.to_csv("new_customers.csv")